### PR TITLE
Plugin Framework outline

### PR DESCRIPTION
**Pull request by Audakel**
Added the abstract interfaces for how I thought a plugin system could work for this project. It handles any type of plugin, not only persistance type plugins.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -60,3 +60,5 @@ java/build/
 java/module_phase_1.xml
 java/testreport/
 testreport/
+.idea/
```
Comment (kupiakos) [inline .gitignore]: These should not be here, as before.

**Comment by kupiakos** [comment]
Shouldn't fc75e4a only be in the `checkstyle` branch?

**Comment by Audakel** [comment]
yeah i removed the style stuff now

File: java/Phase 1.iml [hunk 2]
```diff
@@ -28,5 +28,6 @@
     <orderEntry type="library" name="byte-buddy-1.4.33" level="project" />
     <orderEntry type="library" name="objenesis-2.4" level="project" />
     <orderEntry type="library" name="annotations-13.0" level="project" />
+    <orderEntry type="library" name="checkstyle-7.3-all" level="project" />
```
Comment (kupiakos) [inline java/Phase 1.iml]: Still have the library defined.

File: java/src/server/plugin/IPluginLoader.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+public interface IPluginLoader {
+    /**
+     * Opens the plugin JAR file that the PluginConfig name specified and then
+     * uses an attribute from JAR manifest or the list of all files in the JAR file to find
+     * the class that implements the Plugin interface.
+     * The server will instantiate plugin objects by putting plugin JAR
+     * files on its classpath using URLClassLoader, and then use the Class.forName method to
+     * load the class, and the Class.newInstance method to create them.
+     */
+    ArrayList<IPlugin> loadConfig(ArrayList<PluginConfig> configs);
```
Comment (kupiakos) [inline java/src/server/plugin/IPluginLoader.java]: These should be returning `List`

**Review by kupiakos — CHANGES_REQUESTED** [review]

**Comment by Audakel** [comment]
ok try that

**Comment by Audakel** [comment]
Fixed ArrayList

**Pull request merged**